illumos-utils/src/contract.rs

```diff
@@ -0,0 +1,387 @@
+#[cfg(target_os = "illumos")]
+use libc::ctid_t;
+use slog::{Logger, debug, error};
+use std::ffi::CString;
+use std::ffi::c_char;
+use std::ffi::c_int;
+use std::ffi::c_uint;
+use std::ffi::c_void;
+use std::thread;
+use std::time::Duration;
+
+#[allow(non_camel_case_types)]
+type ct_evthdl_t = *mut c_void;
+#[allow(non_camel_case_types)]
+type ct_evid_t = u64;
+
+#[link(name = "contract")]
+unsafe extern "C" {
```

Comment:
It's probably a bit much, but I wonder about wrapping these all up in a few little wrapper functions so that we can translate the return values into proper `Result` types. I  don't know if it's worth it, though.